### PR TITLE
search frontend: highlight and partition path separators

### PR DESCRIPTION
**Pull request by rvantonder**
This adds highlighting to path separators `/` for repo and file regex values. I hacked on it in December and had time to clean it up now. The idea is that this functionality will also be future-compatible for glob highlighting.

Screen Shot 2021-02-18 at 6 57 36 PM

Screen Shot 2021-02-18 at 6 57 02 PM

It has a nice aesthetic effect of separating the values between `/`, such that hovering on the parts between `/` only show matches for that string (and not the whole thing like `a/b/c`):

https://user-images.githubusercontent.com/888624/108446482-6dfe5500-721b-11eb-990e-26519c2a2510.mp4

Implementation details:

The way it works is to try and first parse a value on `/` (Taking care of escaping). Once we have a list of values of fragments between `/`s, we try to parse those to regular expressions and return regular expression highlights. Doing it the other way around is possible (first regex, then scan `/`) but complicates things, because currently all path-like values are regexp, but not all regexp are path-like. I kind of just chose this route naturally.

The one tricky thing to get right is if we have a regular expression like `foo(/)bar`. Now, parsing on `/` gives you fragments `foo(` and `)bar`, and these fragments are not valid regex independently. But we still want to get regexp highlights for `foo(/)bar`. So this implementation will attempt parsing on `/` first and then highlight regexps, but if it fails, it will parse the entire string as regexp and leave out the `/` processing.

---

I will break up the this file in a subsequent PR, it's too large.

### CODE REVIEW

**Review by rrhyne — APPROVED** [review]
Love it.

File: client/shared/src/search/query/decoratedToken.ts [hunk 1]
```diff
@@ -399,6 +415,107 @@ const mapRegexpMeta = (pattern: Pattern): DecoratedToken[] => {
...
+ * contain path separators like `/`.
+ */
+export const hasPathLikeValue = (field: string): boolean => {
+    const fieldName = field.startsWith('-') ? field.slice(1) : field
+    switch (fieldName.toLocaleLowerCase()) {
+        case 'repo':
+        case 'r':
+        case 'file':
+        case 'f':
+        case 'repohasfile':
```
Comment (lguychard) [inline client/shared/src/search/query/decoratedToken.ts:429]: I vaguely remember discussing this before -- why we don't expose a normalized field name on `filter` tokens to make checks like this more typesafe / less dependent on string manipulation?

**Review by lguychard — APPROVED** [review]

File: client/shared/src/search/query/decoratedToken.ts (hunk 1 above)
Comment (rvantonder) [reply client/shared/src/search/query/decoratedToken.ts:429]: Prior to https://github.com/sourcegraph/sourcegraph/pull/16868, and around the time of interactive search, the filter definitions were kind of in disarray and would require a bit of effort to move around. Things are better now (I think you did the moving, and then I [did some renaming](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/commit/d7abfe5164ff0a7969b9b7cb0f43c7332522963a?visible=7)), so we can revisit this now--I'll spend a couple of hours on this to straighten things out :-)

**Pull request merged**